### PR TITLE
OCPCLOUD-2584: Enable azure infrastructure provider reconcile

### PR DESCRIPTION
**Pull request by RadekManak**
This PR has few changes needed for azure platform:

- Enables reconciliation of AzureCluster. 
- Adds missing backendPool name to hack/template
- E2E test that machine creation for Azure.

### CODE REVIEW

**Review by elmiko — COMMENTED** [review]
/approve
/lgtm

**Comment by nrb** [comment]
I think the assets job will pass once #124 is approved.

The tech preview jobs look like flakes though.

/test e2e-azure-ovn-techpreview

File: e2e/azure_test.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+		Spec: azurev1.AzureClusterSpec{
+			AzureClusterClassSpec: azurev1.AzureClusterClassSpec{
+				Location:         mapiProviderSpec.Location,
+				AzureEnvironment: "AzurePublicCloud",
+				IdentityRef: &corev1.ObjectReference{
+					Name:      clusterName,
+					Namespace: framework.CAPINamespace,
+					Kind:      "AzureClusterIdentity",
+				},
+			},
```
Comment (damdo) [inline e2e/azure_test.go:161]: Is this required for things to work? I was thinking ClusterClass is not a requirement

**Review by damdo — COMMENTED** [review]
Thanks Radek.
Mostly LGTM I have a couple of Qs.

Do we have a presubmit configured that exercises the new E2Es for Azure? 
Basically something similar to `ci/prow/e2e-aws-capi-techpreview` for azure.

If not we should probably add it before merging this PR, so we start getting signal early

**Comment by JoelSpeed** [comment]
I suspect linting is now broken here as I made it stricter recently, sorry
/test lint

**Comment by RadekManak** [comment]
Tested cluster with both PRs and ran E2E on it. The machine was created successfully.

First, we need to merge the provider PR https://github.com/openshift/cluster-api-provider-azure/pull/300. Then this one that I tested manually. Then we can rehearse the https://github.com/openshift/release/pull/54458 presubmits and merge them if it looks good.

**Comment by JoelSpeed** [comment]
> First, we need to merge the provider PR https://github.com/openshift/cluster-api-provider-azure/pull/300. Then this one that I tested manually. Then we can rehearse the https://github.com/openshift/release/pull/54458 presubmits and merge them if it looks good.

If we were to merge the Azure presubmits before the other two, they would fail, which is fine, but should go green on this PR pre-merge right? Just thinking we may not have to do them in that order

**Comment by damdo** [comment]
techpreview jobs are failing due to https://redhat-internal.slack.com/archives/C01CQA76KMX/p1722432194685519 and https://github.com/openshift/ovn-kubernetes/pull/2243#issuecomment-2258686658

We need to rerun these e2es once those are both fixed

File: e2e/go.mod [hunk 2]
```diff
@@ -14,14 +14,21 @@ require (
 	k8s.io/utils v0.0.0-20240102154912-e7106e64919e
 	sigs.k8s.io/cluster-api v1.7.3
 	sigs.k8s.io/cluster-api-provider-aws/v2 v2.5.0
+	sigs.k8s.io/cluster-api-provider-azure v1.15.2
```
Comment (damdo) [inline e2e/go.mod:17]: Should we use https://github.com/kubernetes-sigs/cluster-api-provider-azure/releases/tag/v1.16.0 now that it available upstream and has the `DisableExtensionOperations` available? So we can remove the replace below too! :)

File: e2e/go.mod [hunk 3]
```diff
@@ -14,14 +14,21 @@ require (
...
 	sigs.k8s.io/cluster-api-provider-gcp v1.6.1-0.20240611121101-5a947cf2bf02
 	sigs.k8s.io/cluster-api-provider-ibmcloud v0.8.0
 	sigs.k8s.io/cluster-api-provider-vsphere v1.10.0
 	sigs.k8s.io/controller-runtime v0.17.5
 	sigs.k8s.io/yaml v1.4.0
 )
 
+// Use OpenShift fork to get new APIs before they are released in upstream release.
+replace sigs.k8s.io/cluster-api-provider-azure => github.com/openshift/cluster-api-provider-azure v0.1.0-alpha.3.0.20240626223109-787b95ff79cb
+
```
Comment (damdo) [inline e2e/go.mod:27]: See https://github.com/openshift/cluster-capi-operator/pull/115/files#r1711146863

**Comment by damdo** [comment]
@RadekManak looks like there is a diff in go vendor

**Review by damdo — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**